src/cloudsc_gpu/cloudsc_driver_gpu_scc_cuf_k_caching_mod.F90

```diff
@@ -0,0 +1,362 @@
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFSQSF_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFCQRNG_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFCQSNG_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFSQLTUR_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFSQITUR_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFPLSL_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFPLSN_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFHPSL_d(:, :, :)
+    REAL(KIND=JPRB), ALLOCATABLE, DEVICE :: PFHPSN_d(:, :, :)
+    TYPE(DIM3) :: GRIDDIM, BLOCKDIM
+    
+    ! Device array allocation
+    ALLOCATE (PT_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PQ_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (BUFFER_TMP_d(NPROMA, NLEV, 3 + NCLV, NGPBLKS))
+    ALLOCATE (BUFFER_LOC_d(NPROMA, NLEV, 3 + NCLV, NGPBLKS))
+    ALLOCATE (PVFA_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PVFL_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PVFI_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PDYNA_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PDYNL_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PDYNI_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PHRSW_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PHRLW_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PVERVEL_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PAP_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PAPH_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PLSM_d(NPROMA, NGPBLKS))
+    ALLOCATE (LDCUM_d(NPROMA, NGPBLKS))
+    ALLOCATE (KTYPE_d(NPROMA, NGPBLKS))
+    ALLOCATE (PLU_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PLUDE_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PSNDE_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PMFU_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PMFD_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PA_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PCLV_d(NPROMA, NLEV, NCLV, NGPBLKS))
+    ALLOCATE (PSUPSAT_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PLCRIT_AER_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PICRIT_AER_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PRE_ICE_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PCCN_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PNICE_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PCOVPTOT_d(NPROMA, NLEV, NGPBLKS))
+    ALLOCATE (PRAINFRAC_TOPRFZ_d(NPROMA, NGPBLKS))
+    ALLOCATE (PFSQLF_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFSQIF_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFCQNNG_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFCQLNG_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFSQRF_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFSQSF_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFCQRNG_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFCQSNG_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFSQLTUR_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFSQITUR_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFPLSL_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFPLSN_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFHPSL_d(NPROMA, NLEV + 1, NGPBLKS))
+    ALLOCATE (PFHPSN_d(NPROMA, NLEV + 1, NGPBLKS))
+    
+    ! Copy host to device
```

Comment:
The data offload timers are placed wrong. The outer timer should either sit here, or above the device-side allocation to measure the offload times correctly, no?